File: c/persistent-stack.c
#include <stdio.h>
#include <stdlib.h>

typedef struct PersistentStack PersistentStack;

struct PersistentStack {
    int value;
    int size;
    PersistentStack* parent;
};

PersistentStack createPersistentStack(int value, PersistentStack* parent, int size) {
    PersistentStack stack;
    stack.value = value;
    stack.size = size;
    stack.parent = parent;
    return stack;
}

PersistentStack push(PersistentStack* stack, int value) {
    return createPersistentStack(value, stack, stack->size + 1);
}

PersistentStack pop(PersistentStack* stack) {
    if(stack->parent == 0) {
        printf("stack is empty!\n");
        return *stack;
    }
    return *(stack->parent);
}

int top(PersistentStack * stack) {
    if(stack->parent == 0) {
        printf("stack is empty!\n");
        return 0;
    }
    return stack->value;
}


void printHelper(PersistentStack* stack, int index) {
        if (stack->parent == 0) {
            return;
        }

        if(index > 0) {
            printf(",");
        };
        printf("%d", stack->value);

        printHelper(stack->parent, index + 1);
}


void printStack(PersistentStack* stack, int i) {
    printf("stack%d top: %d, stack: ", i, top(stack));
    printf("[");
    printHelper(stack, 0);
    printf("]\n");
}


int main() {
    int initialSize = 1;
    PersistentStack stack1 = createPersistentStack(0, 0, initialSize);

    PersistentStack stack2 = push(&stack1, 1);
    PersistentStack stack3 = push(&stack2, 2);

    PersistentStack stack4 = pop(&stack3);
    PersistentStack stack5 = pop(&stack4);


    printStack(&stack1, 1);
    printStack(&stack2, 2);
    printStack(&stack3, 3);
    printStack(&stack4, 4);
    printStack(&stack5, 5);

    return 0;
}

File: typescript/persistent-stack.ts

export class PersistentStack<T> {
    private value: T | null = null;
    private parent: PersistentStack<T> | null = null;
    private size: number = 0;
    constructor(value?: any, parent?: PersistentStack<T>, size?: number ) {
        this.value = value ?? null;
        this.parent = parent ?? null;
        this.size = size ?? 0;
    }

    public push(value: T): PersistentStack<T> {
        return new PersistentStack(value, this, this.size + 1)
    };

    public pop(): PersistentStack<T> | null {
        if(this.parent == null) {
            console.log("The stack is empty!")
            return null;
        }
        return this.parent;
    }

    public top(): T | null {
        if(this.parent == null) {
            console.log("The stack is empty!")
            return null;
        }
        return this.value;
    }


    public printStack(): any[] {
        const array = Array.from({length: this.size})
        let node: PersistentStack<T> | null = this;
        for(let i = this.size -1; i >= 0 && node != null; i--) {
            array[i] = node.value;
            node = this.parent;
        }
        return array;
    }
}


const stack1 = new PersistentStack<number>()

const stack2 = stack1.push(1)
const stack3 = stack2.push(2)

const stack4 = stack3.pop();
const stack5 = stack4?.pop();


console.log("stack1 top: " + stack1.top() + " ,stack:" + stack1.printStack())
console.log("stack2 top: " + stack2.top() + " ,stack: " + stack2.printStack())
console.log("stack3 top: " + stack3.top() + " ,stack: " + stack3.printStack())
console.log("stack4 top: " + stack4?.top() + " ,stack: " + stack4?.printStack())
console.log("stack5 top: " + stack5?.top() + " ,stack: " + stack5?.printStack())



